servers/castor/edna.nix

```diff
@@ -0,0 +1,121 @@
+{pkgs, lib, config, ...}:
+  let
+    inherit (builtins) toString typeOf toJSON;
+    inherit (lib) collect concatStringsSep;
+    inherit (pkgs) writeText;
+
+    dns_names = with config.networking;
+      concatStringsSep "," [ "${hostName}.${domain}" "staging.edna.serokell.team" ];
+
+    profile = "/nix/var/nix/profiles/per-user/deploy/edna-docker";
+
+    ports = {
+      backend = {
+        api = 9000;
+      };
+      frontend = {
+        http = 80;
+        https = 443;
+      };
+      prometheus = {
+        node = 9100;
+      };
+    };
+
+    pginit = writeText "init.sql"
+    ''
+      CREATE DATABASE edna;
+    '';
+  in
+  {
+    networking.firewall.allowedTCPPorts = collect (a: (typeOf a) == "int") { inherit (ports) frontend prometheus; };
+
+    virtualisation.docker = {
+      enable = true;
+      logLevel = "warn";
+      storageDriver = "overlay2";
+      volumes = [ "pgdata" ];
+      networks.edna = {};
+    };
+
+    systemd.services.docker-frontend.serviceConfig.RestartSec = "5m";
+
+    virtualisation.oci-containers.containers = let
+      commonOptions = {
+        extraOptions = [
+          # Put all nodes on the same private network
+          "--network=edna"
+
+          # PostgreSQL during initdb
+          "--shm-size=256MB"
+        ];
+      };
+
+    in
+    {
+      frontend = {
+        image = "ghcr.io/serokell/edna-backend";
+        imageFile = "${profile}/ghcr.io-serokell-edna-frontend.tar.gz";
+        dependsOn = [ "backend" ];
+
+        volumes = [
+          "/root/caddy:/.caddy"
+        ];
+
+        ports = with ports.frontend; [
+          "${toString http}:${toString http}"
+          "${toString https}:${toString https}"
+        ];
+
+        environment = {
+          API_HOST = "backend";
+          API_PORT = toString ports.backend.api;
+          DNS_DOMAIN = dns_names;
+        };
+      } // commonOptions;
+
+      backend = let
+        configFile = writeText "config.yaml" (toJSON {
```

Comment:
The format of the config file is quite unstable and the whole concept of this yaml config is debatable, we may even replace it by something else (there were talks about that). For this reason it's quite sad that the config is written in this repo, it means that whenever we update the config structure we have to make a PR to this repo. Is it possible to move more stuff into `edna` repo (here: https://github.com/serokell/edna/pull/11)?

If it doesn't make sense to move anything there, I guess we can live with it.

Reply:
There's a thing or two I could do, but I wouldn't worry too much. I'm aware the config structure is unstable, and I expect that I'll have to change things soon. You can consider the entire construct a placeholder.

This is actually one reason I asked about config merging. It would be ideal for me to only specify options that differ from their defaults.

Reply:
Ok, probably after https://github.com/serokell/edna/pull/8 we will not make config changes for a while (but it's hard to predict), so let's wait for it to be merged, update the config here and stop worrying for now.
AFAIU, if we update the config structure so that the old config no longer can be parsed and merge it, **master** CI build will try to redeploy staging and will fail until we update this place in this repo, right? That's not critical at this stage, just asking.

> This is actually one reason I asked about config merging. It would be ideal for me to only specify options that differ from their defaults.

Yes, I think we will implement config merging and reading from env variables and command line options at some point, it's just a matter of priorities. Currently we have a lot of other work and tight deadlines, so we just provide _some way_ to configure `edna-server` which is not perfect. I'll create an issue about that, but it will stay in backlog for a while.

Reply:
`edna-server` changes have been merged, see https://github.com/serokell/edna/pull/11#issuecomment-799206708

Reply:
> I'll create an issue about that, but it will stay in backlog for a while.

https://issues.serokell.io/issue/EDNA-44